src/repeat.js

```diff
@@ -94,17 +115,74 @@ export class Repeat {
     }
   }
 
-  processItemsByStrategy() {
+  getInnerCollection() {
+    let expression = unwrapExpression(this.sourceExpression);
+    if (!expression) {
+      return null;
+    }
+    return expression.evaluate(this.scope, null);
+  }
+
+  observeInnerCollection() {
+    let items = this.getInnerCollection();
+    if (items instanceof Array) {
+      this.collectionObserver = this.observerLocator.getArrayObserver(items);
+    } else if (items instanceof Map) {
+      this.collectionObserver = this.observerLocator.getMapObserver(items);
+    } else {
+      return false;
+    }
+    this.callContext = 'handleInnerCollectionChanges';
+    this.collectionObserver.subscribe(this.callContext, this);
+    return true;
+  }
+
+  observeCollection() {
     let items = this.items;
     this.collectionObserver = this.collectionStrategy.getCollectionObserver(items);
-    this.collectionStrategy.processItems(items);
     if (this.collectionObserver) {
-      this.callContext = 'handleChanges';
+      this.callContext = 'handleCollectionChanges';
       this.collectionObserver.subscribe(this.callContext, this);
     }
   }
 
-  handleChanges(collection, changes) {
+  processItemsByStrategy() {
+    if (!this.observeInnerCollection()) {
+      this.observeCollection();
+    }
+    this.collectionStrategy.processItems(this.items);
+  }
+
+  handleCollectionChanges(collection, changes) {
     this.collectionStrategy.handleChanges(collection, changes);
   }
+
+  handleInnerCollectionChanges(collection, changes) {
+    let newItems = this.sourceExpression.evaluate(this.scope, this.lookupFunctions);
+    if (newItems === this.items) {
+      return;
+    }
+    this.items = newItems;
+    this.itemsChanged();
+  }
+}
+
+function getSourceExpression(instruction, attrName) {
+  return instruction.behaviorInstructions
+    .filter(bi => bi.originalAttrName === attrName)[0]
+    .attributes
+    .items
+    .sourceExpression;
+}
+
+function unwrapExpression(expression) {
+  let unwrapped = false;
+  while (expression instanceof BindingBehavior) {
+    expression = expression.expression;
+  }
+  while (expression instanceof ValueConverter) {
+    expression = expression.expression;
+    unwrapped = true;
+  }
+  return unwrapped ? expression : null;
 }
```

Comment:
@EisenbergEffect  / @martingust 
Should the `getSourceExpression` and `unwrapExpression` functions go in a separate file?  If so, what should we name it? `expression-utilities.js`?